infobind/cache.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package infobind
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser"
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/parser/terror"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/sqlexec"
+	log "github.com/sirupsen/logrus"
+)
+
+const defaultBindCacheSize = 5
+
+// bindRecord store the basic bind info and bindSql astNode.
+type bindRecord struct {
+	bindMeta
+	ast ast.StmtNode
+}
+
+// bindCache holds a bindDataMap
+//   key: origin sql
+//   value: bindRecord slice.
+type bindCache map[string][]*bindRecord
+
+// Handler hold an atomic bindCache.
+type Handler struct {
+	bind atomic.Value
+}
+
+// BindCacheUpdater use to update the bindCache.
+// BindCacheUpdater is used to update the global bindCache.
+// BindCacheUpdater will update the bind cache pre 3 second in domain gorountine loop. When the tidb server first startup, the updater will load all bind info in memory; then load diff bind info pre 3 second.
+type BindCacheUpdater struct {
+	ctx sessionctx.Context
+
+	parser         *parser.Parser
+	lastUpdateTime types.Time
+	globalHandler  *Handler
+}
+
+type bindMeta struct {
+	OriginalSQL string
+	BindSQL     string
+	Db          string
+	// Status will only be 0 or 1:
+	//   0: bindMeta has been marked as deleted status.
+	//   1: bindMeta is in using status.
+	Status     int64
```

Comment:
how about s/Status/IsValid/?

Reply:
The following pr will add a new state -1. For example we are using a index hint, but the administrator delete the index, the tidb will update the memory status to -1 when the tidb execute sql bind for current sql. Then the following query sql will don't use the sql bind.

Reply:
How about just using strings to specify the status? Integers may confuse users.

Reply:
ok